src/memory/level_up_manager.rs

```diff
@@ -0,0 +1,154 @@
+            4 => LevelUpUpgrade::PhysicalDefense,
+            5 => LevelUpUpgrade::MagicalAttack,
+            6 => LevelUpUpgrade::MagicalDefense,
+            _ => LevelUpUpgrade::None,
+        }
+    }
+}
+
+impl MemoryManagerUpdate for LevelUpManagerData {
+    fn update(
+        &mut self,
+        ctx: &StateContext,
+        manager: &mut UnityMemoryManager,
+    ) -> Result<(), MemoryError> {
+        let memory_context = MemoryContext::create(ctx, manager)?;
+
+        self.update_level_up_screen_active(&memory_context)?;
+        if self.active {
+            self.update_current_character(&memory_context)?;
+            self.update_current_upgrades(&memory_context)?;
+        }
+
+        Ok(())
+    }
+}
+
+impl LevelUpManagerData {
+    // This address checks the onLevelUpDone callback pointer
+    // If this is 0x0, it is not loaded, if it is not 0x0 then its pointing to
+    // a function the game expects to call when the level up screen is over.
+    pub fn update_level_up_screen_active(
+        &mut self,
+        memory_context: &MemoryContext,
+    ) -> Result<(), MemoryError> {
+        if let Ok(active) = memory_context.follow_fields::<u64>(&["onLevelUpDone"]) {
+            self.active = active != 0x0;
+        }
+        Ok(())
+    }
+
+    pub fn update_current_character(
+        &mut self,
+        memory_context: &MemoryContext,
+    ) -> Result<(), MemoryError> {
+        if let Ok(current_character_ptr) = memory_context.follow_fields::<u64>(&[
+            "currentCharacter",
+            "stateMachine",
+            "currentState",
+            "player",
+            "characterDefinitionId",
+        ]) {
+            let current_character_str =
+                memory_context.read_pointer::<ArrayWString<16>>(current_character_ptr + 0x14)?;
+
+            let current_character = match String::from_utf16(current_character_str.as_slice()) {
+                Ok(value) => value,
+                Err(_) => {
+                    return Ok(());
+                }
+            };
+            self.current_character = PlayerPartyCharacter::parse(&current_character)
+        }
+        Ok(())
+    }
+
+    pub fn update_current_upgrades(
+        &mut self,
+        memory_context: &MemoryContext,
+    ) -> Result<(), MemoryError> {
+        if let Ok(upgrades) = memory_context.follow_fields::<u64>(&["currentLevelUpUpgrades"]) {
+            let current_upgrades =
+                UnityList::<LevelUpUpgradeButton>::read(memory_context.process, upgrades)?;
+            self.current_upgrades = current_upgrades;
+            // convert to enumerator and take first with selected true and return its index
+            let active = self
+                .current_upgrades
+                .items
+                .iter()
+                .enumerate()
+                .find(|(_i, u)| u.selected)
+                .map(|(i, _u)| i)
+                .take();
+            self.upgrade_index = active.unwrap() as u32;
```

Comment:
When it's unwrapped like this it almost looks nicer/more readable to just have a for loop looking through it XD

Reply:
Yea... iterators optimize well or something.